subsys/bluetooth/audio/csis_client.c

```diff
@@ -140,6 +143,110 @@ static struct bt_csis *lookup_instance_by_set_info(const struct bt_csis_client_s
 }
 
+static int member_rank_compare_asc(const void *m1, const void *m2)
+{
+	struct bt_csis_client_set_member *member_1 = *(struct bt_csis_client_set_member **)m1;
+	struct bt_csis_client_set_member *member_2 = *(struct bt_csis_client_set_member **)m2;
+	struct bt_csis *inst_1;
+	struct bt_csis *inst_2;
+
+	/* TODO: lookup_instance_by_set_info should be able to work on a const
+	 * member, but it relies on `bt_conn_index` which doesn't take a const
+	 * bt_conn.
+	 */
+
+	inst_1 = lookup_instance_by_set_info(member_1, active.info);
+	inst_2 = lookup_instance_by_set_info(member_2, active.info);
+
+	if (inst_1 == NULL) {
+		BT_ERR("inst_1 was NULL for member %p", member_1);
+		return 0;
+	}
+
+	if (inst_2 == NULL) {
+		BT_ERR("inst_2 was NULL for member %p", member_2);
+		return 0;
+	}
+
+	if (inst_1->cli.rank < inst_2->cli.rank) {
+		return -1;
+	} else if (inst_1->cli.rank > inst_2->cli.rank) {
+		return 1;
+	}
+
+	return 0;
+}
+
+static int member_rank_compare_desc(const void *m1, const void *m2)
+{
+	struct bt_csis_client_set_member *member_1 = *(struct bt_csis_client_set_member **)m1;
+	struct bt_csis_client_set_member *member_2 = *(struct bt_csis_client_set_member **)m2;
+	struct bt_csis *inst_1;
+	struct bt_csis *inst_2;
+
+	/* TODO: lookup_instance_by_set_info should be able to work on a const
+	 * member, but it relies on `bt_conn_index` which doesn't take a const
+	 * bt_conn.
+	 */
+
+	inst_1 = lookup_instance_by_set_info(member_1, active.info);
+	inst_2 = lookup_instance_by_set_info(member_2, active.info);
+
+	if (inst_1 == NULL) {
+		BT_ERR("inst_1 was NULL for member %p", member_1);
+		return 0;
+	}
+
+	if (inst_2 == NULL) {
+		BT_ERR("inst_2 was NULL for member %p", member_2);
+		return 0;
+	}
+
+	if (inst_1->cli.rank < inst_2->cli.rank) {
+		return 1;
+	} else if (inst_1->cli.rank > inst_2->cli.rank) {
+		return -1;
+	}
+
+	return 0;
+}
+
+static void active_members_store_ordered(struct bt_csis_client_set_member *members[],
+					 size_t count,
+					 const struct bt_csis_client_set_info *info,
+					 bool ascending)
+{
+	(void)memcpy(active.members, members, count * sizeof(members[0]));
+	active.members_count = count;
+	active.info = info;
+
+	qsort(active.members, count, sizeof(members[0]),
+	      ascending ? member_rank_compare_asc : member_rank_compare_desc);
+
+	if (IS_ENABLED(CONFIG_ASSERT) &&
+	    IS_ENABLED(CONFIG_BT_DEBUG_CSIS_CLIENT)) {
+
+		for (size_t i = 1U; i < count; i++) {
+			const struct bt_csis *inst_1 =
+				lookup_instance_by_set_info(active.members[i - 1], info);
+			const struct bt_csis *inst_2 =
+				lookup_instance_by_set_info(active.members[i], info);
+			const uint8_t rank_1 = inst_1->cli.rank;
+			const uint8_t rank_2 = inst_2->cli.rank;
+
+			if (ascending) {
+				__ASSERT(rank_1 <= rank_2,
+					 "Members not sorted by ascending rank %u - %u",
+					 rank_1, rank_2);
+			} else {
+				__ASSERT(rank_1 >= rank_2,
+					"Members not sorted by descending rank %u - %u",
+					rank_1, rank_2);
```

Comment:
Is this assert needed? Couldn't be verified by a bsim_test or unit test?

Reply:
The application does have access to the rank value at this point, so it won't be able to do such as verification (actually tried to implement this in the babblesim tests first time around).

Reply:
My point is that you sort the members, and then try to assert if not sorted. I'm OK with asserting things that one is not 100% certain, but here...

Reply:
The solutions I see are
1) (current) Do an assert to verify that the sorting functionality works
2) Have a babblesim test use the internal header file and access the rank
3) Expose the rank to the application
4) Replace the asserts with a `BT_ERR`

Reply:
If the check is necessary, I guess the assert is the best option from those you listed